rust/src/tx_builder/tx_inputs_builder.rs

```diff
@@ -0,0 +1,369 @@
+            script: script.clone(),
+            datum: datum.clone(),
+            redeemer: redeemer.clone(),
+        }
+    }
+
+    pub fn script(&self) -> PlutusScript {
+        self.script.clone()
+    }
+
+    pub fn datum(&self) -> PlutusData {
+        self.datum.clone()
+    }
+
+    pub fn redeemer(&self) -> Redeemer {
+        self.redeemer.clone()
+    }
+
+    fn clone_with_redeemer_index(&self, index: &BigNum) -> Self {
+        Self {
+            script: self.script.clone(),
+            datum: self.datum.clone(),
+            redeemer: self.redeemer.clone_with_index(index),
+        }
+    }
+}
+
+
+#[wasm_bindgen]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct PlutusWitnesses(pub(crate) Vec<PlutusWitness>);
+
+#[wasm_bindgen]
+impl PlutusWitnesses {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn get(&self, index: usize) -> PlutusWitness {
+        self.0[index].clone()
+    }
+
+    pub fn add(&mut self, elem: &PlutusWitness) {
+        self.0.push(elem.clone());
+    }
+
+    pub(crate) fn collect(&self) -> (PlutusScripts, PlutusList, Redeemers) {
+        let mut s = PlutusScripts::new();
+        let mut d = PlutusList::new();
+        let mut r = Redeemers::new();
+        self.0.iter().for_each(|w| {
+            s.add(&w.script);
+            d.add(&w.datum);
+            r.add(&w.redeemer);
+        });
+        (s, d, r)
+    }
+}
+
+impl From<Vec<PlutusWitness>> for PlutusWitnesses {
+    fn from(scripts: Vec<PlutusWitness>) -> Self {
+        scripts.iter().fold(PlutusWitnesses::new(), |mut scripts, s| {
+            scripts.add(s);
+            scripts
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum ScriptWitnessType {
+    NativeScriptWitness(NativeScript),
+    PlutusScriptWitness(PlutusWitness),
+}
+
+// We need to know how many of each type of witness will be in the transaction so we can calculate the tx fee
+#[derive(Clone, Debug)]
+pub struct MockWitnessSet {
+    vkeys: RequiredSignersSet,
+    scripts: LinkedHashMap<ScriptHash, Option<ScriptWitnessType>>,
+    bootstraps: BTreeSet<Vec<u8>>,
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct TxInputsBuilder {
+    inputs: Vec<(TxBuilderInput, Option<ScriptHash>)>,
+    input_types: MockWitnessSet,
+}
+
+pub(crate) fn get_bootstraps(inputs: &TxInputsBuilder) -> BTreeSet<Vec<u8>> {
+    inputs.input_types.bootstraps.clone()
+}
+
+#[wasm_bindgen]
+impl TxInputsBuilder {
+
+    pub fn new() -> Self {
+        Self {
+            inputs: Vec::new(),
+            input_types: MockWitnessSet {
+                vkeys: BTreeSet::new(),
+                scripts: LinkedHashMap::new(),
+                bootstraps: BTreeSet::new(),
+            },
+        }
+    }
+
+    fn push_input(&mut self, e: (TxBuilderInput, Option<ScriptHash>)) {
+        self.inputs.push(e);
+        self.inputs.sort_by(|(a, _), (b, _)| a.input.cmp(&b.input));
```

Comment:
If we have sort for each push, can we use smth like a BTreeMap or  binary heap ? 
https://doc.rust-lang.org/std/collections/struct.BTreeMap.html
https://doc.rust-lang.org/stable/std/collections/struct.BinaryHeap.html

Reply:
good point, will check

Reply:
done